test/test_rtc.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+
+Copyright (c) 2020, Arvid Norberg
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/config.hpp"
+#include <libtorrent/torrent.hpp>
+#include <libtorrent/magnet_uri.hpp>
+
+#include "test.hpp"
+#include "test_utils.hpp"
+#include "setup_transfer.hpp"
+#include "settings.hpp"
+#include "session_mock.hpp"
+
+#if TORRENT_USE_RTC
+
+#include <boost/asio.hpp>
+
+#include <iostream>
+#include <chrono>
+#include <numeric>
+#include <algorithm>
+#include <random>
+
+#include <cstdio>
+#include <cstdarg>
+
+using namespace std::placeholders;
+using namespace std::chrono_literals;
+using namespace lt;
+
+using aux::rtc_signaling;
+using aux::rtc_stream;
+using aux::rtc_stream_init;
+using aux::rtc_offer;
+using aux::rtc_answer;
+
+namespace {
+
+boost::asio::io_context io_context;
+
+session_mock ses1(io_context);
+torrent tor1(ses1, false, parse_magnet_uri("magnet:?xt=urn:btih:cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd"));
+
+session_mock ses2(io_context);
+torrent tor2(ses2, false, parse_magnet_uri("magnet:?xt=urn:btih:cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd"));
+
+bool success = false;
+
+void run_test() {
+	success = false;
+
+	auto begin_time = clock_type::now();
+	auto end_time = begin_time + 60s;
+	do {
+		io_context.run_one_until(end_time);
+		io_context.restart();
+	}
+	while (!success && clock_type::now() < end_time);
+
+	ses1.print_alerts(begin_time);
+	ses1.print_alerts(begin_time);
+
+	TEST_CHECK(success);
```

Comment:
when running this on Linux, this test fails for me. I haven't had a chance to look into the details of why it's failing yet. There isn't much output from the test giving hints.